Nanostack factory configuration support

artokin opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Read key and configuration values from KCM during startup.
Use mbed-cloud-client availability to detect if KCM is available th
the build. If KCM is not available, use Mbed OS default values.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Nanostack factory configuration support

Read key and configuration values from KCM during startup.
Use mbed-cloud-client availability to detect if KCM is available th
the build. If KCM is not available, use Mbed OS default values.

mikter commented on features/nanostack/mbed-mesh-api/source/MeshKcm.cpp:
```diff
@@ -0,0 +1,93 @@
+    return KCM_STATUS_ERROR;
+}
+
+void mesh_kcm_wisun_network_name_init(char *network_name_buf, size_t network_name_buf_len)
```
This is read which returns configured value if KCM not available

Commit:
Factory configuration support for network-size

mikaleppanen commented on features/nanostack/mbed-mesh-api/source/MeshKcm.cpp:
```diff
@@ -0,0 +1,140 @@
+            config_status = 0;
+        }
+    } else {
+        strncpy(network_name_buf, MBED_CONF_MBED_MESH_API_WISUN_NETWORK_NAME, network_name_buf_len);
```
is the "#ifdef MBED_CONF_MBED_MESH_API_WISUN_NETWORK_NAME" needed here?

artokin replied:
By default Wi-SUN network name is always configured. Why we should make it optional?

mikaleppanen replied:
Ok, looks like there is default for that in mesh api mbed_lib.json... could it still be possible that application overrides that with NULL definition so that the define would not be generated? And then it is later set by function interface?

artokin replied:
fixed.

Commit:
Update flag used to detect FCC availability

Use MBED_CONF_MBED_CLOUD_CLIENT_EXTERNAL_SST_SUPPORT instead of
ARM_UC_PROFILE_MBED_CLOUD_CLIENT to detect FCC availability.

Commit:
Factory config support for device regulatory_domain

Add factory configuration support for regulatory_domain,
operating_mode and operating_class.

Commit:
Factory config support radius server

Add factory configuration support for radius server address and
shared secret.

mikter commented on features/nanostack/mbed-mesh-api/source/WisunInterface.cpp:
```diff
@@ -119,13 +131,18 @@ nsapi_error_t WisunInterface::configure()
-    status = set_network_size(MBED_CONF_MBED_MESH_API_WISUN_NETWORK_SIZE);
-    if (status != MESH_ERROR_NONE) {
-        tr_error("Failed to set network size");
+    ret_val = mesh_kcm_wisun_network_size_init(&network_size);
```
Previously it was Ok not to have configuration as it would use stack defaults you haven changed that, but I think it should be working like previously

artokin replied:
Ok, I'll change that back to original.
So if there is a configuration error in KCM data, we use Mbed OS defaults. If Mbed OS default does not exists, we use Nanostack defaults.

Commit:
Review corrections

-Fallback to Mbed OS config if KCM config is not available
-Add compilation flags around network name configuration

teetak01 reviewed (changes requested):
We are reconsidering the design, so I would hold this PR for now.
